lib/Serialization/SerializedModuleLoader.cpp

```diff
@@ -44,6 +106,132 @@ SerializedModuleLoader::~SerializedModuleLoader() = default;
 MemoryBufferSerializedModuleLoader::~MemoryBufferSerializedModuleLoader() =
     default;
 
+void SerializedModuleLoaderBase::collectVisibleTopLevelModuleNamesImpl(
+    SmallVectorImpl<Identifier> &names, StringRef extension) const {
+  llvm::SmallString<16> moduleSuffix;
+  moduleSuffix += '.';
+  moduleSuffix += file_types::getExtension(file_types::TY_SwiftModuleFile);
+
+  llvm::SmallString<16> suffix;
+  suffix += '.';
+  suffix += extension;
+
+  SmallVector<SmallString<64>, 2> targetFiles;
+  forEachTargetModuleBasename(Ctx, [&](StringRef targetName) {
+    targetFiles.emplace_back(targetName);
+    targetFiles.back() += suffix;
+  });
+
+  auto &fs = *Ctx.SourceMgr.getFileSystem();
+
+  // Apply \p body for each directory entry in \p dirPath.
+  auto forEachDirectoryEntryPath =
+      [&](StringRef dirPath, llvm::function_ref<void(StringRef)> body) {
+        std::error_code errorCode;
+        llvm::vfs::directory_iterator DI = fs.dir_begin(dirPath, errorCode);
+        llvm::vfs::directory_iterator End;
+        for (; !errorCode && DI != End; DI.increment(errorCode))
+          body(DI->path());
+      };
+
+  // Check whether target specific module file exists or not in given directory.
+  // $PATH/{arch}.{extension}
+  auto checkTargetFiles = [&](StringRef path) -> bool {
```

Comment:
Is there any point in checking these instead of just detecting the presence of the .swiftmodule directory/file?  We aren't checking if the module is importable.

Reply:
That'd be faster but might mean less common code between both code paths.

Reply:
Good point. I'm okay with this if it's not a big perf issue and keeps the code simpler.